File: 医药翻译-编码目录/frontend/src/api/file.ts
import { request } from '@/api/client'
import type { ApiEnvelope } from '@/types'

export const FILE_CHUNK_SIZE = 5 * 1024 * 1024

export async function checkFile(payload: {
  fileHash: string
  filename: string
  fileSize: number
  totalChunks: number
}) {
  const response = await request<ApiEnvelope<{ exists: boolean; file_id?: string | null }>>('/api/file/check', {
    method: 'POST',
    body: JSON.stringify({
      file_hash: payload.fileHash,
      filename: payload.filename,
      file_size: payload.fileSize,
      total_chunks: payload.totalChunks,
    }),
  })

  return {
    exists: response.data.exists,
    fileId: response.data.file_id ?? undefined,
  }
}

export async function uploadChunk(payload: {
  fileHash: string
  chunkIndex: number
  totalChunks: number
  chunk: Blob
  filename: string
}) {
  const form = new FormData()
  form.append('file_hash', payload.fileHash)
  form.append('chunk_index', String(payload.chunkIndex))
  form.append('total_chunks', String(payload.totalChunks))
  form.append('chunk', payload.chunk, payload.filename)

  return request<ApiEnvelope<{ success: boolean }>>('/api/file/chunk', {
    method: 'POST',
    body: form,
  })
}

export async function mergeChunks(payload: {
  fileHash: string
  filename: string
  totalChunks: number
  mimeType: string
}) {
  const response = await request<ApiEnvelope<{ file_id: string }>>('/api/file/merge', {
    method: 'POST',
    body: JSON.stringify({
      file_hash: payload.fileHash,
      filename: payload.filename,
      total_chunks: payload.totalChunks,
      mime_type: payload.mimeType,
    }),
  })
  return { fileId: response.data.file_id }
}

File: 医药翻译-编码目录/backend/app/routers/file.py
from fastapi import APIRouter, Depends, File, Form, UploadFile
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db_session
from app.deps import get_current_user
from app.models.user import User
from app.schemas.common import ApiResponse
from app.schemas.file import FileCheckRequest, FileCheckResponse, FileMergeRequest, FileMergeResponse
from app.services.file_service import file_service


router = APIRouter(prefix='/file', tags=['file'])


@router.post('/check', response_model=ApiResponse[FileCheckResponse])
async def check_file(
    payload: FileCheckRequest,
    session: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[FileCheckResponse]:
    return ApiResponse(data=await file_service.check_file(session, current_user, payload))


@router.post('/chunk', response_model=ApiResponse[dict])
async def upload_chunk(
    file_hash: str = Form(...),
    chunk_index: int = Form(...),
    total_chunks: int = Form(...),
    chunk: UploadFile = File(...),
    session: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[dict]:
    await file_service.save_chunk(session, current_user, file_hash, chunk_index, total_chunks, chunk)
    return ApiResponse(data={'success': True, 'chunk_index': chunk_index, 'total_chunks': total_chunks})


@router.post('/merge', response_model=ApiResponse[FileMergeResponse])
async def merge_file(
    payload: FileMergeRequest,
    session: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[FileMergeResponse]:
    return ApiResponse(data=await file_service.merge_chunks(session, current_user, payload))
